### PR TITLE
JH2 - 5/10/2016

### PR DESCRIPTION
**Pull request by olegbaslak**
done lesson1.task04

### CODE REVIEW

File: src/com/rct/homework/lesson1/task04/Validator.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Validates String arguments
+ *
+ * @author Oleg Baslak
+ * @version 1.0
+ * @since 30-09-2016
+ */
+
+class Validator {
+
+    static double[] parse(String[] args) {
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/Validator.java]: Missed comment for method parse

File: src/com/rct/homework/lesson1/task04/Validator.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        double[] parsedCoefficients = new double[Main.NUMBER_OF_ARGUMENTS];
+        try {
+            for (int i = 0; i < parsedCoefficients.length; i++) {
+                parsedCoefficients[i] = Double.parseDouble(args[i]);
+            }
+        }catch (NumberFormatException e){
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/Validator.java]: Missed spaces

File: src/com/rct/homework/lesson1/task04/Validator.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            e.getMessage();
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/Validator.java]: Unused return value of statement e.getMessage();

**Review by AlexanderSoroka — CHANGES_REQUESTED** [review]
It looks quite good but there are some issues to be fixed

File: src/com/rct/homework/lesson1/task04/Validator.java (hunk 3 above)
Comment (olegbaslak) [reply src/com/rct/homework/lesson1/task04/Validator.java]: Check new commit

File: src/com/rct/homework/lesson1/task04/Main.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     * @param args command line arguments
+     */
+    public static void main(String[] args) {
+        if (args.length != NUMBER_OF_ARGUMENTS) {
+            System.out.println(WRONG_INPUT_MESSAGE);
+            System.exit(1);
+        }
+
+        try {
+            SquareEquation squareEquation = new SquareEquation(Validator.parse(args));
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/Main.java:35]: Is zero acceptable value for paramA? I don't see this check

File: src/com/rct/homework/lesson1/task04/Validator.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    static double[] parse(String[] args) {
+        double[] parsedCoefficients = new double[Main.NUMBER_OF_ARGUMENTS];
+        try {
+            for (int i = 0; i < parsedCoefficients.length; i++) {
+                parsedCoefficients[i] = Double.parseDouble(args[i]);
+            }
+        } catch (NumberFormatException e) {
+            System.out.println(PARSE_ERROR);
+
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/Validator.java]: Useless blank line

File: src/com/rct/homework/lesson1/task05/Main.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Entrance point to the app
+     *
+     * @param args command line arguments
+     * @throws TriangleNotExistsException if triangle with inputted lengths
+     *                                    don't exist
+     */
+    public static void main(String[] args) {
+
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task05/Main.java]: Useless blank line

File: src/com/rct/homework/lesson1/task05/Main.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        System.out.println(INPUT_REQUSET);
+
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task05/Main.java]: Useless blank line

**Review by AlexanderSoroka — CHANGES_REQUESTED** [review]

File: src/com/rct/homework/lesson1/task05/Main.java (hunk 7 above)
Comment (olegbaslak) [reply src/com/rct/homework/lesson1/task05/Main.java]: I tried to divide code to logical blocks

File: src/com/rct/homework/lesson1/task04/Main.java (hunk 4 above)
Comment (olegbaslak) [reply src/com/rct/homework/lesson1/task04/Main.java:35]: Oooops

**Comment by olegbaslak** [comment]
Check new commits please

File: src/com/rct/homework/lesson1/task05/InputController.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ * @since 03-10-2016
+ */
+class InputController {
+
+    /* Message string constants */
+    private static final String NAN_ERROR_MESSAGE = "is not a number";
+    private static final String TRY_AGAIN_MESSAGE = "Try again";
+
+    /**
+     * Gets {@code} paramsNumber arguments from console input.
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task05/InputController.java]: This doesn;t say that the method ignores negatives

File: src/com/rct/homework/lesson1/task05/InputController.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+        Scanner scanner = new Scanner(System.in);
+
+        String readString = "";
+        double readNumber;
+
+        int index = 0;
+        while (index < paramsNumber) {
+            try {
+                readString = scanner.next();
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task05/InputController.java]: What if user inputs something very big like 2e2048?

**Review by AlexanderSoroka — CHANGES_REQUESTED** [review]

**Comment by olegbaslak** [comment]
Fresh chaneges incoming

File: src/com/rct/homework/lesson1/task04/SquareEquation.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Find discriminant of equation
+     *
+     * @return counted discriminant
+     */
+    private double findDiscriminant() {
+        double d;
+        d = this.paramB * this.paramB - 4 * this.paramA * this.paramC;
+        return d;
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/SquareEquation.java]: return paramB_paramB - 4_paramA*paramC; is shorter and simpler to read, isn't it?

File: src/com/rct/homework/lesson1/task04/SquareEquation.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+     *
+     * @return array with counted root
+     * @throws ComplexRootsOnlyException if equation has complex only roots
+     * @throws NoRootsException          if only {@code c} param presented, so there no methods
+     *                                   to find roots
+     */
+    double[] findRoots() throws ComplexRootsOnlyException, NoRootsException {
+        double[] roots = new double[2];
+
+        /* eq. c = 0 */
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/SquareEquation.java]: Here and everywhere use // comments

File: src/com/rct/homework/lesson1/task04/SquareEquation.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        if (Validator.isZero(paramA) && Validator.isZero(paramB)) {
+            throw new NoRootsException();
+        }
+
+        /* eq. bx = -c */
+        if (Validator.isZero(paramA) && !Validator.isZero(paramB)) {
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/SquareEquation.java]: This is not quadratic equation

Comment (AlexanderSoroka) [reply src/com/rct/homework/lesson1/task04/SquareEquation.java]: I think most of these checks are useless

**Review by AlexanderSoroka — CHANGES_REQUESTED** [review]

File: src/com/rct/homework/lesson1/task04/Validator.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+     * @param args arguments from command line to parse
+     * @return parsed coefficients
+     */
+    static double[] parse(String[] args) {
+        double[] parsedCoefficients = new double[Main.NUMBER_OF_ARGUMENTS];
+        try {
+            for (int i = 0; i < parsedCoefficients.length; i++) {
+                parsedCoefficients[i] = Double.parseDouble(args[i]);
+            }
+        } catch (NumberFormatException e) {
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task04/Validator.java]: Is it better to expose it for caller and let it decides if application should stop or something else? Method that hides inside System.exit(1) is really bad idea - at least it must be well-documented

File: src/com/rct/homework/lesson1/task05/Main.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+     *
+     * @param args command line arguments
+     * @throws TriangleNotExistsException if triangle with inputted lengths
+     *                                    don't exist
+     */
+    public static void main(String[] args) {
+        System.out.println(INPUT_REQUSET);
+        InputController inputController = new InputController();
+        double[] sides = new double[0];
+        try {
```
Comment (AlexanderSoroka) [inline src/com/rct/homework/lesson1/task05/Main.java]: Let combine this 2 try catch to one

**Review by AlexanderSoroka — CHANGES_REQUESTED** [review]

**Review by AlexanderSoroka — APPROVED** [review]

**Pull request merged**